File: proj_python/simulation.py
"""Task 4 — Simulation and validation for the VSA term project.

Uses the learned policy, transition model, and penalty probabilities to
generate synthetic interaction sequences with the **same** sequence structure
(length and starting ``phi_src``) as the real dataset, then compares:

    * action distribution
    * penalty rates (global and per action)
    * state visitation (``phi_src`` and ``phi_dst`` marginals)

For each marginal, reports **absolute probability differences**
``|P_data − P_sim|`` (per category / per action / per state as appropriate).

Two run modes
-------------
Direct (reads ``policy.csv``, ``penalty.csv``, ``transition__*.csv`` from the
Task~1 output directory and the real CSV for empirical marginals; writes
under ``./outputs/task4_simulation/``; also prints):

    python simulation.py [--csv PATH] [--policy-dir DIR] [--out DIR]

Imported by ``load_all()`` (passes ``EstimationResult`` in memory; still loads
the real CSV from ``csv_path`` for empirical marginals; also prints):

    from simulation import run_simulation
    result = run_simulation(estimation=..., csv_path=...)
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Tuple, Union

import numpy as np
import pandas as pd

from util import (
    CLMN_ALPHA,
    CLMN_BETA,
    CLMN_PHI_DST,
    CLMN_PHI_SRC,
    CLMN_SEQ_ID,
    CLMN_STEP,
    DEFAULT_CSV,
    EstimationResult,
    HERE,
    load_vsa_dataset,
)

DEFAULT_POLICY_DIR = HERE / "outputs" / "task1_estimation"
DEFAULT_OUT = HERE / "outputs" / "task4_simulation"

PENALTY_FILE = "penalty.csv"
POLICY_FILE = "policy.csv"


@dataclass
class SimulationResult:
    """Structured return value of ``run_simulation``."""

    n_synthetic_rows: int
    abs_diff_action: pd.Series
    abs_diff_phi_src: pd.Series
    abs_diff_phi_dst: pd.Series
    abs_diff_penalty_per_action: pd.Series
    abs_diff_penalty_global: float
    l1_action: float
    l1_phi_src: float
    l1_phi_dst: float
    l1_penalty_per_action: float
    synthetic_df: pd.DataFrame
    source: str
    random_seed: int


def _section(title: str) -> str:
    bar = "=" * 72
    return f"\n{bar}\n{title}\n{bar}"


def _sample_from_row(
    rng: np.random.Generator,
    labels: List[str],
    probs: np.ndarray,
) -> str:
    """Categorical sample; uniform if ``probs`` is all-zero."""
    probs = np.asarray(probs, dtype=float)
    s = probs.sum()
    if s < 1e-15:
        probs = np.ones(len(labels), dtype=float) / len(labels)
    else:
        probs = probs / s
    return str(rng.choice(labels, p=probs))


def _load_model_from_dir(policy_dir: Path) -> Tuple[pd.DataFrame, Dict[str, pd.DataFrame], pd.Series]:
    p_path = policy_dir / POLICY_FILE
    pen_path = policy_dir / PENALTY_FILE
    if not p_path.exists():
        raise FileNotFoundError(f"Expected {p_path} (Task 1 output). Run estimation.py first.")
    if not pen_path.exists():
        raise FileNotFoundError(f"Expected {pen_path} (Task 1 output). Run estimation.py first.")

    policy = pd.read_csv(p_path, index_col=0, encoding="utf-8")
    policy.index = policy.index.astype(str)
    policy.columns = policy.columns.astype(str)

    pen_df = pd.read_csv(pen_path, index_col=0, encoding="utf-8")
    if "c_k" not in pen_df.columns:
        raise ValueError(f"{pen_path} must contain column 'c_k'.")
    penalty = pen_df["c_k"].astype(float)
    penalty.index = penalty.index.astype(str)
    penalty.name = "c_k"

    transition: Dict[str, pd.DataFrame] = {}
    for path in sorted(policy_dir.glob("transition__*.csv")):
        action = path.stem[len("transition__") :]
        tab = pd.read_csv(path, index_col=0, encoding="utf-8")
        tab.index = tab.index.astype(str)
        tab.columns = tab.columns.astype(str)
        transition[action] = tab

    missing = [a for a in policy.columns if a not in transition]
    if missing:
        raise ValueError(
            f"Missing transition file(s) for action(s) {missing} under {policy_dir}. "
            f"Found transition keys: {list(transition.keys())}."
        )
    return policy, transition, penalty


def _simulate_block(
    real_df: pd.DataFrame,
    policy: pd.DataFrame,
    transition: Dict[str, pd.DataFrame],
    penalty: pd.Series,
    rng: np.random.Generator,
) -> pd.DataFrame:
    """Generate synthetic rows matching real (sequence_id, step) layout."""
    meta = (
        real_df.sort_values([CLMN_SEQ_ID, CLMN_STEP])
        .groupby(CLMN_SEQ_ID, sort=False)
        .agg(length=(CLMN_STEP, "count"), phi0=(CLMN_PHI_SRC, "first"))
    )
    lengths = meta["length"].astype(int).to_numpy()
    phi0s = meta["phi0"].astype(str).to_numpy()
    seq_ids = meta.index.to_numpy()

    states = list(policy.index.astype(str))
    actions = list(policy.columns.astype(str))

    rows: List[dict] = []
    for sid, L, phi in zip(seq_ids, lengths, phi0s):
        phi = str(phi)
        if phi not in policy.index:
            phi = states[0]
        for step in range(1, int(L) + 1):
            prow = policy.loc[phi].to_numpy(dtype=float)
            a = _sample_from_row(rng, actions, prow)
            tmat = transition[a]
            labels = list(tmat.columns.astype(str))
            if phi in tmat.index:
                trow = tmat.loc[phi].to_numpy(dtype=float)
            else:
                trow = np.ones(len(labels), dtype=float) / len(labels)
            phi_next = _sample_from_row(rng, labels, trow)
            p_pen = float(penalty.reindex([a]).fillna(0.0).iloc[0])
            p_pen = min(max(p_pen, 0.0), 1.0)
            beta = int(rng.binomial(1, p_pen))
            rows.append(
                {
                    CLMN_SEQ_ID: int(sid),
                    CLMN_STEP: step,
                    CLMN_PHI_SRC: phi,
                    CLMN_PHI_DST: phi_next,
                    CLMN_ALPHA: a,
                    CLMN_BETA: beta,
                }
            )
            phi = phi_next

    return pd.DataFrame(rows)


def _marginal_action(df: pd.DataFrame) -> pd.Series:
    s = df[CLMN_ALPHA].value_counts(normalize=True, sort=False)
    s.name = "P"
    return s


def _marginal_state(df: pd.DataFrame, col: str) -> pd.Series:
    s = df[col].value_counts(normalize=True, sort=False)
    s.name = "P"
    return s


def _penalty_per_action(df: pd.DataFrame) -> pd.Series:
    return df.groupby(CLMN_ALPHA)[CLMN_BETA].mean().sort_index()


def _aligned_abs_diff(p_data: pd.Series, p_sim: pd.Series) -> pd.Series:
    idx = sorted(set(p_data.index.astype(str)) | set(p_sim.index.astype(str)))
    d = p_data.reindex(idx, fill_value=0.0).astype(float)
    s = p_sim.reindex(idx, fill_value=0.0).astype(float)
    out = (d - s).abs()
    out.name = "abs_diff"
    return out


def run_simulation(
    estimation: Optional[EstimationResult] = None,
    *,
    csv_path: Union[Path, str, None] = None,
    policy_dir: Optional[Union[Path, str]] = None,
    output_dir: Optional[Union[Path, str]] = None,
    print_to_console: bool = True,
    random_seed: int = 42,
) -> SimulationResult:
    """Generate synthetic sequences and compare marginals to the dataset.

    Parameters
    ----------
    estimation
        Task~1 result from ``load_all()``; if ``None``, load model matrices from
        ``policy_dir``.
    csv_path
        Real dataset CSV for empirical marginals (defaults to ``util.DEFAULT_CSV``).
    policy_dir
        Task~1 output folder when ``estimation`` is ``None``.
    """
    csv_path = Path(csv_path or DEFAULT_CSV)
    real_df = load_vsa_dataset(csv_path)

    if estimation is not None:
        policy = estimation.policy
        transition = estimation.transition
        penalty = estimation.penalty
        source = "estimation-result"
    else:
        pdir = Path(policy_dir) if policy_dir is not None else DEFAULT_POLICY_DIR
        policy, transition, penalty = _load_model_from_dir(pdir)
        source = f"files: {pdir}"

    rng = np.random.default_rng(random_seed)
    sim_df = _simulate_block(real_df, policy, transition, penalty, rng)

    P_a_d = _marginal_action(real_df)
    P_a_s = _marginal_action(sim_df)
    abs_a = _aligned_abs_diff(P_a_d, P_a_s)

    P_ps_d = _marginal_state(real_df, CLMN_PHI_SRC)
    P_ps_s = _marginal_state(sim_df, CLMN_PHI_SRC)
    abs_ps = _aligned_abs_diff(P_ps_d, P_ps_s)

    P_pd_d = _marginal_state(real_df, CLMN_PHI_DST)
    P_pd_s = _marginal_state(sim_df, CLMN_PHI_DST)
    abs_pd = _aligned_abs_diff(P_pd_d, P_pd_s)

    pen_d = _penalty_per_action(real_df)
    pen_s = _penalty_per_action(sim_df)
    abs_pen_a = _aligned_abs_diff(pen_d, pen_s)

    g_d = float(real_df[CLMN_BETA].mean())
    g_s = float(sim_df[CLMN_BETA].mean())
    abs_g = abs(g_d - g_s)

    l1_a = float(abs_a.sum())
    l1_ps = float(abs_ps.sum())
    l1_pd = float(abs_pd.sum())
    l1_pen_a = float(abs_pen_a.sum())

    result = SimulationResult(
        n_synthetic_rows=len(sim_df),
        abs_diff_action=abs_a,
        abs_diff_phi_src=abs_ps,
        abs_diff_phi_dst=abs_pd,
        abs_diff_penalty_per_action=abs_pen_a,
        abs_diff_penalty_global=abs_g,
        l1_action=l1_a,
        l1_phi_src=l1_ps,
        l1_phi_dst=l1_pd,
        l1_penalty_per_action=l1_pen_a,
        synthetic_df=sim_df,
        source=source,
        random_seed=random_seed,
    )

    lines: List[str] = []
    lines.append(_section(f"Task 4 — Simulation   (source: {source}, seed={random_seed})"))
    lines.append(f"Real rows: {len(real_df)}  |  Synthetic rows: {len(sim_df)}")
    lines.append("")
    lines.append("|P_data − P_sim| — action distribution")
    lines.append(abs_a.round(6).to_string())
    lines.append(f"  L1 sum of abs diffs (actions): {l1_a:.6f}")
    lines.append("")
    lines.append("|P_data − P_sim| — phi_src visitation")
    lines.append(abs_ps.round(6).to_string())
    lines.append(f"  L1 sum: {l1_ps:.6f}")
    lines.append("")
    lines.append("|P_data − P_sim| — phi_dst visitation")
    lines.append(abs_pd.round(6).to_string())
    lines.append(f"  L1 sum: {l1_pd:.6f}")
    lines.append("")
    lines.append("|mean_beta_data − mean_beta_sim| per action (penalty rate)")
    lines.append(abs_pen_a.round(6).to_string())
    lines.append(f"  L1 sum (per-action |Δ|): {l1_pen_a:.6f}")
    lines.append("")
    lines.append(f"Global penalty rate  |Δ|:  data={g_d:.6f}  sim={g_s:.6f}  |Δ|={abs_g:.6f}")

    report = "\n".join(lines)
    if print_to_console:
        print(report)

    if output_dir is not None:
        out = Path(output_dir)
        out.mkdir(parents=True, exist_ok=True)

        pd.concat(
            [
                P_a_d.rename("P_data"),
                P_a_s.rename("P_sim"),
                abs_a.rename("abs_diff"),
            ],
            axis=1,
        ).to_csv(out / "marginal_action.csv", encoding="utf-8")

        pd.concat(
            [
                P_ps_d.rename("P_data"),
                P_ps_s.rename("P_sim"),
                abs_ps.rename("abs_diff"),
            ],
            axis=1,
        ).to_csv(out / "marginal_phi_src.csv", encoding="utf-8")

        pd.concat(
            [
                P_pd_d.rename("P_data"),
                P_pd_s.rename("P_sim"),
                abs_pd.rename("abs_diff"),
            ],
            axis=1,
        ).to_csv(out / "marginal_phi_dst.csv", encoding="utf-8")

        pd.concat(
            [
                pen_d.rename("penalty_rate_data"),
                pen_s.rename("penalty_rate_sim"),
                abs_pen_a.rename("abs_diff"),
            ],
            axis=1,
        ).to_csv(out / "penalty_rate_per_action.csv", encoding="utf-8")

        pd.DataFrame(
            [
                {
                    "abs_diff_penalty_global": abs_g,
                    "l1_action": l1_a,
                    "l1_phi_src": l1_ps,
                    "l1_phi_dst": l1_pd,
                    "l1_penalty_per_action": l1_pen_a,
                    "random_seed": random_seed,
                }
            ]
        ).to_csv(out / "simulation_summary.csv", index=False, encoding="utf-8")

        sim_df.to_csv(out / "synthetic_sequences.csv", index=False, encoding="utf-8")
        (out / "report.txt").write_text(report, encoding="utf-8")
        if print_to_console:
            print(f"\nWrote outputs to: {out}")

    return result


def _parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Task 4 — Simulation (VSA term project)")
    p.add_argument("--csv", type=Path, default=DEFAULT_CSV, help="Real dataset CSV")
    p.add_argument(
        "--policy-dir",
        type=Path,
        default=DEFAULT_POLICY_DIR,
        help="Task 1 output directory (default: %(default)s)",
    )
    p.add_argument("--out", type=Path, default=DEFAULT_OUT, help="Output directory")
    p.add_argument("--seed", type=int, default=42, help="RNG seed")
    return p.parse_args()


if __name__ == "__main__":
    args = _parse_args()
    run_simulation(
        estimation=None,
        csv_path=args.csv,
        policy_dir=args.policy_dir,
        output_dir=args.out,
        print_to_console=True,
        random_seed=args.seed,
    )

File: proj_java/src/main/java/cse8003/vsa/appl/Simulation.java
package cse8003.vsa.appl;

import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.stream.Collectors;

/** Task 4 simulation (Python {@code simulation.py}). */
public final class Simulation {

    public record SimulationResult(
            int nSyntheticRows,
            Map<String, Double> absDiffAction,
            Map<String, Double> absDiffPhiSrc,
            Map<String, Double> absDiffPhiDst,
            Map<String, Double> absDiffPenaltyPerAction,
            double absDiffPenaltyGlobal,
            double l1Action,
            double l1PhiSrc,
            double l1PhiDst,
            List<Util.VsaRow> syntheticRows,
            String source,
            int randomSeed) {}

    private Simulation() {}

    public static SimulationResult runSimulation(
            Util.EstimationResult estimation,
            Path csvPath,
            boolean printToConsole,
            int randomSeed) throws IOException {

        List<Util.VsaRow> realRows = Util.loadVsaDataset(csvPath);
        Random rng = new Random(randomSeed);
        List<Util.VsaRow> simRows = simulateBlock(realRows, estimation, rng);

        Map<String, Double> pActionData = marginalAction(realRows);
        Map<String, Double> pActionSim = marginalAction(simRows);
        Map<String, Double> absAction = alignedAbsDiff(pActionData, pActionSim);

        Map<String, Double> pSrcData = marginalState(realRows, true);
        Map<String, Double> pSrcSim = marginalState(simRows, true);
        Map<String, Double> absSrc = alignedAbsDiff(pSrcData, pSrcSim);

        Map<String, Double> pDstData = marginalState(realRows, false);
        Map<String, Double> pDstSim = marginalState(simRows, false);
        Map<String, Double> absDst = alignedAbsDiff(pDstData, pDstSim);

        Map<String, Double> penData = Util.meanBetaByAction(realRows);
        Map<String, Double> penSim = Util.meanBetaByAction(simRows);
        Map<String, Double> absPen = alignedAbsDiff(penData, penSim);

        double gData = realRows.stream().mapToInt(Util.VsaRow::beta).average().orElse(0.0);
        double gSim = simRows.stream().mapToInt(Util.VsaRow::beta).average().orElse(0.0);
        double absG = Math.abs(gData - gSim);

        double l1A = absAction.values().stream().mapToDouble(Double::doubleValue).sum();
        double l1Src = absSrc.values().stream().mapToDouble(Double::doubleValue).sum();
        double l1Dst = absDst.values().stream().mapToDouble(Double::doubleValue).sum();

        SimulationResult result = new SimulationResult(
                simRows.size(),
                absAction,
                absSrc,
                absDst,
                absPen,
                absG,
                l1A,
                l1Src,
                l1Dst,
                simRows,
                "estimation-result",
                randomSeed);

        if (printToConsole) {
            printReport(result, realRows.size(), gData, gSim);
        }
        return result;
    }

    private static List<Util.VsaRow> simulateBlock(
            List<Util.VsaRow> realRows, Util.EstimationResult model, Random rng) {

        record SeqMeta(int seqId, int length, String phi0) {}

        Map<Integer, SeqMeta> meta = new LinkedHashMap<>();
        realRows.stream()
                .sorted(Comparator.comparingInt(Util.VsaRow::sequenceId).thenComparingInt(Util.VsaRow::step))
                .forEach(r -> {
                    if (!meta.containsKey(r.sequenceId())) {
                        meta.put(r.sequenceId(), new SeqMeta(r.sequenceId(), 0, r.phiSrc()));
                    }
                });
        Map<Integer, Long> lengths = realRows.stream()
                .collect(Collectors.groupingBy(Util.VsaRow::sequenceId, Collectors.counting()));
        for (var e : lengths.entrySet()) {
            SeqMeta old = meta.get(e.getKey());
            meta.put(e.getKey(), new SeqMeta(old.seqId(), e.getValue().intValue(), old.phi0()));
        }

        List<String> states = model.states();
        List<String> actions = model.actions();
        List<Util.VsaRow> out = new ArrayList<>();

        for (SeqMeta seq : meta.values()) {
            String phi = seq.phi0();
            if (!model.policy().rowLabels().contains(phi)) {
                phi = states.get(0);
            }
            for (int step = 1; step <= seq.length(); step++) {
                String a = sampleFromRow(rng, actions, model.policy(), phi);
                Util.Table2D tmat = model.transition().get(a);
                List<String> dstLabels = tmat.colLabels();
                String phiNext;
                if (tmat.rowLabels().contains(phi)) {
                    phiNext = sampleFromRow(rng, dstLabels, tmat, phi);
                } else {
                    double[] uniform = new double[dstLabels.size()];
                    double p = 1.0 / dstLabels.size();
                    for (int i = 0; i < uniform.length; i++) {
                        uniform[i] = p;
                    }
                    phiNext = sampleFromProbs(rng, dstLabels, uniform);
                }
                double pPen = Math.min(1.0, Math.max(0.0, model.penalty().getOrDefault(a, 0.0)));
                int beta = rng.nextDouble() < pPen ? 1 : 0;
                out.add(new Util.VsaRow(seq.seqId(), step, phi, phiNext, a, beta));
                phi = phiNext;
            }
        }
        return out;
    }

    private static String sampleFromRow(
            Random rng, List<String> labels, Util.Table2D table, String row) {
        double[] probs = new double[labels.size()];
        for (int i = 0; i < labels.size(); i++) {
            probs[i] = table.get(row, labels.get(i));
        }
        return sampleFromProbs(rng, labels, probs);
    }

    private static String sampleFromProbs(Random rng, List<String> labels, double[] probs) {
        double sum = 0.0;
        for (double p : probs) {
            sum += p;
        }
        if (sum < 1e-15) {
            double u = 1.0 / labels.size();
            for (int i = 0; i < probs.length; i++) {
                probs[i] = u;
            }
        } else {
            for (int i = 0; i < probs.length; i++) {
                probs[i] /= sum;
            }
        }
        double u = rng.nextDouble();
        double c = 0.0;
        for (int i = 0; i < labels.size(); i++) {
            c += probs[i];
            if (u < c) {
                return labels.get(i);
            }
        }
        return labels.get(labels.size() - 1);
    }

    private static Map<String, Double> marginalAction(List<Util.VsaRow> rows) {
        Map<String, Long> counts = rows.stream()
                .collect(Collectors.groupingBy(Util.VsaRow::alpha, Collectors.counting()));
        long total = rows.size();
        Map<String, Double> out = new LinkedHashMap<>();
        counts.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(e -> out.put(e.getKey(), (double) e.getValue() / total));
        return out;
    }

    private static Map<String, Double> marginalState(List<Util.VsaRow> rows, boolean src) {
        Map<String, Long> counts = rows.stream()
                .collect(Collectors.groupingBy(r -> src ? r.phiSrc() : r.phiDst(), Collectors.counting()));
        long total = rows.size();
        Map<String, Double> out = new LinkedHashMap<>();
        counts.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(e -> out.put(e.getKey(), (double) e.getValue() / total));
        return out;
    }

    private static Map<String, Double> alignedAbsDiff(Map<String, Double> data, Map<String, Double> sim) {
        LinkedHashSet<String> keys = new LinkedHashSet<>();
        keys.addAll(data.keySet());
        keys.addAll(sim.keySet());
        Map<String, Double> out = new LinkedHashMap<>();
        for (String k : keys.stream().sorted().collect(Collectors.toList())) {
            out.put(k, Math.abs(data.getOrDefault(k, 0.0) - sim.getOrDefault(k, 0.0)));
        }
        return out;
    }

    private static void printReport(SimulationResult r, int realRows, double gData, double gSim) {
        StringBuilder lines = new StringBuilder();
        lines.append(Util.section("Task 4 — Simulation   (source: estimation-result, seed=" + r.randomSeed() + ")"));
        lines.append("Real rows: ").append(realRows).append("  |  Synthetic rows: ").append(r.nSyntheticRows()).append("\n\n");
        lines.append("|P_data − P_sim| — action distribution\n");
        lines.append(Util.formatSeries(Util.CLMN_ALPHA, r.absDiffAction())).append("\n");
        lines.append(String.format("  L1 sum of abs diffs (actions): %.6f%n%n", r.l1Action()));
        lines.append("|P_data − P_sim| — phi_src visitation\n");
        lines.append(Util.formatSeries(Util.CLMN_PHI_SRC, r.absDiffPhiSrc())).append("\n");
        lines.append(String.format("  L1 sum: %.6f%n%n", r.l1PhiSrc()));
        lines.append("|P_data − P_sim| — phi_dst visitation\n");
        lines.append(Util.formatSeries(Util.CLMN_PHI_DST, r.absDiffPhiDst())).append("\n");
        lines.append(String.format("  L1 sum: %.6f%n%n", r.l1PhiDst()));
        lines.append("|mean_beta_data − mean_beta_sim| per action (penalty rate)\n");
        lines.append(Util.formatSeries(Util.CLMN_ALPHA, r.absDiffPenaltyPerAction())).append("\n");
        double l1Pen = r.absDiffPenaltyPerAction().values().stream().mapToDouble(Double::doubleValue).sum();
        lines.append(String.format("  L1 sum (per-action |Δ|): %.6f%n%n", l1Pen));
        lines.append(String.format("Global penalty rate  |Δ|:  data=%.6f  sim=%.6f  |Δ|=%.6f%n", gData, gSim, r.absDiffPenaltyGlobal()));
        System.out.print(lines);
    }

}
